pkg/compose/create.go

```diff
@@ -858,24 +858,73 @@ MOUNTS:
 			// `Mount` is preferred but does not offer option to created host path if missing
 			// so `Bind` API is used here with raw volume string
 			// see https://github.com/moby/moby/issues/43483
-			for _, v := range service.Volumes {
-				if v.Target == m.Target {
-					switch {
-					case string(m.Type) != v.Type:
-						v.Source = m.Source
-						fallthrough
-					case !requireMountAPI(v.Bind):
-						binds = append(binds, v.String())
-						continue MOUNTS
-					}
-				}
+			v := findVolumeByTarget(service.Volumes, m.Target)
+			if v == nil {
+				continue MOUNTS
+			}
+			switch {
+			case v.Type != types.VolumeTypeBind:
+				v.Source = m.Source
+				fallthrough
+			case !requireMountAPI(v.Bind):
+				vol := findVolumeByName(p.Volumes, m.Source)
+				binds = append(binds, toBindString(vol.Name, v))
+				continue MOUNTS
+			}
+		}
+		if m.Type == mount.TypeVolume {
+			v := findVolumeByTarget(service.Volumes, m.Target)
+			if v == nil {
+				continue MOUNTS
+			}
+			vol := findVolumeByName(p.Volumes, m.Source)
+			if vol.Driver == "local" && vol.DriverOpts["o"] == "bind" {
```

Comment:
I think the trick also requires `--opt type=none`

Reply:
This hack is also supported for long by compose (https://github.com/docker/compose/blob/v1/compose/config/config.py#L494-L498) and relies on `o=bind` and `device` option being set